File: lib/models/Order.ts
import mongoose from 'mongoose';

export interface IOrderItem {
  menuItemId: mongoose.Types.ObjectId;
  name: string;
  price: number;
  quantity: number;
  isVeg: boolean;
  platform: 'swiggy' | 'zomato' | 'ondc';
}

export interface IOrder extends mongoose.Document {
  userId: mongoose.Types.ObjectId;
  restaurantId: mongoose.Types.ObjectId;
  restaurantName: string;
  items: IOrderItem[];
  subtotal: number; // Before discounts
  totalAmount: number; // After discounts
  deliveryFee: number;
  // Coupon and gift card
  coupon?: {
    code: string;
    couponId: string;
    discountAmount: number;
    deliveryDiscount: number;
  };
  giftCard?: {
    code: string;
    giftCardId: string;
    amountUsed: number;
  };
  deliveryAddress: {
    fullAddress: string;
    landmark?: string;
    city: string;
    pincode: string;
    phone: string;
  };
  status: 'pending' | 'confirmed' | 'preparing' | 'out_for_delivery' | 'delivered' | 'cancelled';
  platform: 'swiggy' | 'zomato' | 'ondc';
  paymentStatus: 'pending' | 'processing' | 'completed' | 'failed';
  paymentMethod?: 'cod' | 'online';
  paymentDetails?: {
    razorpayOrderId?: string;
    razorpayPaymentId?: string;
    razorpaySignature?: string;
  };
  paymentFailureReason?: string;
  specialInstructions?: string;
  // Commission tracking
  commission: {
    rate: number; // Percentage (10-15%)
    amount: number; // Calculated commission amount
    status: 'pending' | 'paid' | 'refunded';
    paidAt?: Date;
  };
  restaurantPayout: {
    amount: number; // totalAmount - commission.amount
    status: 'pending' | 'paid' | 'refunded';
    paidAt?: Date;
  };
  createdAt: Date;
  updatedAt: Date;
}

const orderItemSchema = new mongoose.Schema({
  menuItemId: { type: mongoose.Schema.Types.ObjectId, required: true },
  name: { type: String, required: true },
  price: { type: Number, required: true },
  quantity: { type: Number, required: true, min: 1 },
  isVeg: { type: Boolean, required: true },
  platform: { type: String, enum: ['swiggy', 'zomato', 'ondc'], required: true }
});

const orderSchema = new mongoose.Schema({
  userId: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'User',
    required: true,
    index: true
  },
  restaurantId: { 
    type: mongoose.Schema.Types.ObjectId, 
    ref: 'Restaurant',
    required: true,
    index: true
  },
  restaurantName: { type: String, required: true },
  items: [orderItemSchema],
  subtotal: { type: Number, required: true, min: 0 },
  totalAmount: { type: Number, required: true, min: 0 },
  deliveryFee: { type: Number, required: true, default: 0 },
  // Coupon and gift card
  coupon: {
    code: { type: String, uppercase: true },
    couponId: { type: String },
    discountAmount: { type: Number, default: 0 },
    deliveryDiscount: { type: Number, default: 0 }
  },
  giftCard: {
    code: { type: String, uppercase: true },
    giftCardId: { type: String },
    amountUsed: { type: Number, default: 0 }
  },
  deliveryAddress: {
    fullAddress: { type: String, required: true },
    landmark: { type: String },
    city: { type: String, required: true },
    pincode: { type: String, required: true },
    phone: { type: String, required: true }
  },
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'preparing', 'out_for_delivery', 'delivered', 'cancelled'],
    default: 'pending',
    index: true
  },
  platform: {
    type: String,
    enum: ['swiggy', 'zomato', 'ondc'],
    required: true
  },
  paymentStatus: {
    type: String,
    enum: ['pending', 'processing', 'completed', 'failed'],
    default: 'pending'
  },
  paymentMethod: {
    type: String,
    enum: ['cod', 'online']
  },
  paymentDetails: {
    razorpayOrderId: { type: String },
    razorpayPaymentId: { type: String },
    razorpaySignature: { type: String }
  },
  paymentFailureReason: { type: String },
  specialInstructions: { type: String },
  // Commission tracking
  commission: {
    rate: { type: Number, required: true, default: 12 }, // 12% default
    amount: { type: Number, required: true, default: 0 },
    status: { 
      type: String, 
      enum: ['pending', 'paid', 'refunded'],
      default: 'pending'
    },
    paidAt: { type: Date }
  },
  restaurantPayout: {
    amount: { type: Number, required: true, default: 0 },
    status: { 
      type: String, 
      enum: ['pending', 'paid', 'refunded'],
      default: 'pending'
    },
    paidAt: { type: Date }
  }
}, {
  timestamps: true
});

// Index for efficient queries
orderSchema.index({ userId: 1, createdAt: -1 });
orderSchema.index({ restaurantId: 1, createdAt: -1 });

// Calculate commission and restaurant payout before saving
orderSchema.pre('save', function() {
  if (this.isNew || this.isModified('totalAmount') || this.isModified('commission.rate')) {
    const commissionRate = this.commission?.rate || 12; // Default 12%
    const commissionAmount = (this.totalAmount * commissionRate) / 100;
    const restaurantPayoutAmount = this.totalAmount - commissionAmount;

    this.commission = {
      rate: commissionRate,
      amount: Math.round(commissionAmount * 100) / 100, // Round to 2 decimals
      status: this.commission?.status || 'pending',
      paidAt: this.commission?.paidAt
    };

    this.restaurantPayout = {
      amount: Math.round(restaurantPayoutAmount * 100) / 100, // Round to 2 decimals
      status: this.restaurantPayout?.status || 'pending',
      paidAt: this.restaurantPayout?.paidAt
    };
  }
});

export const Order = mongoose.models.Order || mongoose.model<IOrder>('Order', orderSchema);

File: mobile/lib/models/order.dart
class Order {
  final String id;
  final String orderNumber;
  final String restaurantName;
  final List<OrderItem> items;
  final double totalAmount;
  final String status;
  final String paymentStatus;
  final String? paymentMethod;
  final DeliveryAddress deliveryAddress;
  final String platform;
  final String? specialInstructions;
  final DateTime createdAt;
  final DateTime updatedAt;

  Order({
    required this.id,
    required this.orderNumber,
    required this.restaurantName,
    required this.items,
    required this.totalAmount,
    required this.status,
    required this.paymentStatus,
    this.paymentMethod,
    required this.deliveryAddress,
    required this.platform,
    this.specialInstructions,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'] as String,
      orderNumber: json['orderNumber'] as String,
      restaurantName: json['restaurantName'] as String,
      items: (json['items'] as List)
          .map((item) => OrderItem.fromJson(item as Map<String, dynamic>))
          .toList(),
      totalAmount: (json['totalAmount'] as num).toDouble(),
      status: json['status'] as String,
      paymentStatus: json['paymentStatus'] as String,
      paymentMethod: json['paymentMethod'] as String?,
      deliveryAddress: DeliveryAddress.fromJson(json['deliveryAddress'] as Map<String, dynamic>),
      platform: json['platform'] as String,
      specialInstructions: json['specialInstructions'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  String getStatusDisplay() {
    switch (status) {
      case 'pending':
        return 'Order Placed';
      case 'confirmed':
        return 'Confirmed';
      case 'preparing':
        return 'Being Prepared';
      case 'out_for_delivery':
        return 'Out for Delivery';
      case 'delivered':
        return 'Delivered';
      case 'cancelled':
        return 'Cancelled';
      default:
        return status;
    }
  }

  String getPaymentStatusDisplay() {
    switch (paymentStatus) {
      case 'pending':
        return 'Payment Pending';
      case 'completed':
        return 'Paid';
      case 'failed':
        return 'Payment Failed';
      default:
        return paymentStatus;
    }
  }
}

class OrderItem {
  final String menuItemId;
  final String name;
  final double price;
  final int quantity;
  final bool isVeg;
  final String platform;

  OrderItem({
    required this.menuItemId,
    required this.name,
    required this.price,
    required this.quantity,
    required this.isVeg,
    required this.platform,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      menuItemId: json['menuItemId'] as String,
      name: json['name'] as String,
      price: (json['price'] as num).toDouble(),
      quantity: json['quantity'] as int,
      isVeg: json['isVeg'] as bool,
      platform: json['platform'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'menuItemId': menuItemId,
      'name': name,
      'price': price,
      'quantity': quantity,
      'isVeg': isVeg,
      'platform': platform,
    };
  }

  double get totalPrice => price * quantity;
}

class DeliveryAddress {
  final String fullAddress;
  final String? landmark;
  final String city;
  final String pincode;
  final String phone;

  DeliveryAddress({
    required this.fullAddress,
    this.landmark,
    required this.city,
    required this.pincode,
    required this.phone,
  });

  factory DeliveryAddress.fromJson(Map<String, dynamic> json) {
    return DeliveryAddress(
      fullAddress: json['fullAddress'] as String,
      landmark: json['landmark'] as String?,
      city: json['city'] as String,
      pincode: json['pincode'] as String,
      phone: json['phone'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'fullAddress': fullAddress,
      'landmark': landmark,
      'city': city,
      'pincode': pincode,
      'phone': phone,
    };
  }
}
